### PR TITLE
Some more fixes

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
Today's harvest from working on the elliptic surfaces.

### CODE REVIEW

File: experimental/DoubleAndHyperComplexes/src/Morphisms/simplified_complexes.jl [hunk 1]
```diff
@@ -228,7 +228,7 @@ function (fac::SimplifiedChainFactory)(d::AbsHyperComplex, Ind::Tuple)
       w = Tinv[i]
       new_entries = Vector{Tuple{Int, elem_type(base_ring(w))}}()
       for (real_j, b) in w
-        j = findfirst(==(real_j), J)
+        j = findfirst(k->k==real_j, J)
```
Comment (lgoettgens) [inline experimental/DoubleAndHyperComplexes/src/Morphisms/simplified_complexes.jl]:
```suggestion
        j = findfirst(==(real_j), J)
```
why was this changed back (compared to https://github.com/oscar-system/Oscar.jl/pull/4126)?

File: src/Rings/mpoly-ideals.jl [hunk 2]
```diff
@@ -11,7 +11,7 @@ Given a vector `V` of polynomials in `R`, return the ideal of `R` generated by t
 
 # Examples
 ```jldoctest
-julia> R, (x, y) = polynomial_ring(QQ, [:x, :y])
+julia> R, (x, y) = polynomial_ring(QQ, ["x", "y"])
```
Comment (lgoettgens) [inline src/Rings/mpoly-ideals.jl]: this one and the similar changes should not get changed back (compared to https://github.com/oscar-system/Oscar.jl/pull/4133).

File: src/Rings/mpoly-ideals.jl [hunk 3]
```diff
@@ -642,11 +654,8 @@ Return whether `I` is a radical ideal.
 Computes the radical.
 """
 @attr Bool function is_radical(I::MPolyIdeal)
-  if has_attribute(I, :is_prime) && is_prime(I)
-    return true
-  end
-
-  return I == radical(I)
+  has_attribute(I, :is_prime) && return is_prime(I)
```
Comment (lgoettgens) [inline src/Rings/mpoly-ideals.jl]: this change looks wrong to my eyes.
Consider a non-prime, radical ideal. This new version would return `false`, although the ideal is indeed radical

File: src/Rings/mpoly-ideals.jl [hunk 4]
```diff
@@ -898,31 +907,26 @@ Multivariate polynomial ring in 2 variables over number field graded by
   y -> [1]
 ```
 """
-@attr Vector{Tuple{typeof(I), typeof(I), ideal_type(change_base_ring(rationals_as_number_field()[1], base_ring(I))[1]), Int64}} function absolute_primary_decomposition(I::MPolyIdeal{<:MPolyRingElem{QQFieldElem}})
```
Comment (lgoettgens) [inline src/Rings/mpoly-ideals.jl]: all of the below has been changed by https://github.com/oscar-system/Oscar.jl/pull/4071. Why is it changed back?

**Comment by thofma** [comment]
I think a lot of changes from #4133 and #4126 are undone here. I guess this was not on purpose.

**Review by lgoettgens — DISMISSED** [review]

**Comment by HechtiDerLachs** [comment]
Yes, I'm already working on it. It's a bit annoying, because I had to work on the compute-server today, but there I can't put my login-credentials for GH. Thus, I manually copied the files via scp and this is what you get...

File: src/Rings/mpoly-ideals.jl (hunk 3 above)
Comment (HechtiDerLachs) [reply src/Rings/mpoly-ideals.jl]: True. Thanks for the catch.

File: src/Rings/mpoly-ideals.jl [hunk 5]
```diff
@@ -642,12 +654,10 @@ Return whether `I` is a radical ideal.
 Computes the radical.
 """
 @attr Bool function is_radical(I::MPolyIdeal)
-  if has_attribute(I, :is_prime) && is_prime(I)
-    return true
-  end
-
-  return I == radical(I)
+  has_attribute(I, :is_prime) && return is_prime(I)
```
Comment (HechtiDerLachs) [inline src/Rings/mpoly-ideals.jl]:
```suggestion
  has_attribute(I, :is_prime) && is_prime(I) && return true
```

File: src/Rings/mpoly-ideals.jl [hunk 6]
```diff
@@ -2249,3 +2260,10 @@ function flag_pluecker_ideal(ring::MPolyRing{<: FieldElem}, dimensions::Vector{I
                    isReduced=true,
                    isGB=true))
 end
+
+# Since most ideals implement `==`, they have to implement the hash function.
+# See issue #4143 for problems entailed. This should fix it. 
+function hash(I::Ideal, c::UInt)
+  return UInt(0)
+end
```
Comment (HechtiDerLachs) [inline src/Rings/mpoly-ideals.jl]: Any objections? @lgoettgens , @fingolfin ?

File: src/Rings/mpoly-ideals.jl [hunk 7]
```diff
@@ -1149,7 +1170,14 @@ function minimal_primes(
       end
       J = K
     end
-    result = unique!(filter!(!is_one, vcat([minimal_primes(j; algorithm, factor_generators=false) for j in J]...)))
+    # unique! seems to fail here. We have to do it manually.
+    pre_result = filter!(!is_one, vcat([minimal_primes(j; algorithm, factor_generators=false) for j in J]...))
```
Comment (lgoettgens) [inline src/Rings/mpoly-ideals.jl:1206]:
```suggestion
    pre_result = filter!(!is_one, reduce(vcat, [minimal_primes(j; algorithm, factor_generators=false) for j in J]))
```
save some allocations

File: src/Rings/mpoly-ideals.jl [hunk 8]
```diff
@@ -2249,3 +2277,11 @@ function flag_pluecker_ideal(ring::MPolyRing{<: FieldElem}, dimensions::Vector{I
                    isReduced=true,
                    isGB=true))
 end
+
+# Since most ideals implement `==`, they have to implement the hash function.
+# See issue #4143 for problems entailed. Interestingly, this does not yet fix 
+# the failure of unique! on lists of ideals.
+function hash(I::Ideal, c::UInt)
+  return UInt(0)
+end
```
Comment (lgoettgens) [inline src/Rings/mpoly-ideals.jl]:
```suggestion
function hash(I::Ideal, h::UInt)
  return h
end
```
returning the second value is save and a small optimisation.

Comment (benlorenz) [reply src/Rings/mpoly-ideals.jl]: Hashing the concrete type and maybe even the parent ring should also be fast and help avoid a few comparisons.

File: src/Rings/mpoly-ideals.jl [hunk 9]
```diff
@@ -2249,3 +2278,11 @@ function flag_pluecker_ideal(ring::MPolyRing{<: FieldElem}, dimensions::Vector{I
                    isReduced=true,
                    isGB=true))
 end
+
+# Since most ideals implement `==`, they have to implement the hash function.
+# See issue #4143 for problems entailed. Interestingly, this does not yet fix 
+# the failure of unique! on lists of ideals.
+function hash(I::Ideal, c::UInt)
+  return hash(typeof(I), hash(base_ring(I), c))
+end
```
Comment (HechtiDerLachs) [inline src/Rings/mpoly-ideals.jl]: I advocate for this solution now. It is a matter of fact that ideals can not be hashed in general and even for polynomial rings this is probably expensive if not impossible (we need to choose an ordering which might be a dead end for a specific ideal). Yet, the `==` function is vital to how we use ideals and we need to be able to use them in vectors/lists/dictionaries.

Comment (simonbrandhorst) [reply src/Rings/mpoly-ideals.jl]: Is it safe to hash the type? After all two mathematically equal objects may have different types. 
(Thinking e.g. of IdealSheaf, RadicalIdealSheaf, PrimeIdealSheaf etc.)

Comment (lgoettgens) [reply src/Rings/mpoly-ideals.jl]: > Is it safe to hash the type? After all two mathematically equal objects may have different types. (Thinking e.g. of IdealSheaf, RadicalIdealSheaf, PrimeIdealSheaf etc.)

It is only safe if you don't implement `==` between this type and other types. In your example, you shouldn't include the type in hashes.

Comment (HechtiDerLachs) [reply src/Rings/mpoly-ideals.jl]: Ok, good to know. I was only following the suggestion by @benlorenz . But you're probably right.

**Comment by fieker** [comment]
On Fri, Sep 27, 2024 at 01:30:28AM -0700, Simon Brandhorst wrote:
> @simonbrandhorst commented on this pull request.
> 
> 
> 
> > +# Since most ideals implement `==`, they have to implement the hash function.
> +# See issue #4143 for problems entailed. Interestingly, this does not yet fix 
> +# the failure of unique! on lists of ideals.
> +function hash(I::Ideal, c::UInt)
> +  return hash(typeof(I), hash(base_ring(I), c))
> +end
> 
> Is it safe to hash the type? After all two mathematically equal objects may have different types. 
> (Thinking e.g. of IdealSheaf, RadicalIdealSheaf, PrimeIdealSheaf etc.)
Tough one: Julia tries/ tried to make sure any number comparing equal
have the same hash. We don't - way too expensive.

Is there any situation where you want a dict or a set where the keys/
elements are different types and then catch that 2 objects of different
types are "the same"?

**Comment by HechtiDerLachs** [comment]
> Is there any situation where you want a dict or a set where the keys/ elements are different types and then catch that 2 objects of different types are "the same"?

Probably yes. For instance, it would make sense to introduce a lazy type for radical ideals, say `RadicalIdeal`. This would have a special dispatch for ideal membership so that the computation of the radical can be avoided. Yet, comparison with the standard types of ideals for the respective ring could give that two ideals are equal, even though their type differs.

**Review by simonbrandhorst — APPROVED** [review]

File: src/Rings/mpoly-ideals.jl [hunk 10]
```diff
@@ -572,7 +571,31 @@ Ideal generated by
...
+      S, iso, iso_inv = simplify(Q)
+      is_zero(ngens(S)) && return I # This only happens when all variables can be eliminated. 
+      # Then the ideal defines a reduced point.
+      J = modulus(S)
+      pre_res = _compute_radical(J)
+      pre_res = ideal(R, elem_type(R)[lift(iso_inv(S(g))) for g in gens(pre_res)])
+      res = ideal(R, small_generating_set(pre_res + I))
+      set_attribute!(res, :is_radical=>true)
+      return res
+    end
```
Comment (afkafkafk13) [inline src/Rings/mpoly-ideals.jl:616]: As you specifically asked my opinion on this code block: Yes, this definitely makes sense, as it prepares the objects better for the use of Singular's primdec.lib.

File: test/Rings/MPolyAnyMap/MPolyRing.jl [hunk 11]
```diff
@@ -124,7 +124,7 @@
     Kxz, (z1, z2) = Kx[:z1, :z2];
     f = hom(Kxz, Kxz, hom(Kx, Kxz, [z1, z2]), [z2, z1])
     g = hom(Kxz, Kxz, hom(Kx, Kx, [y, x]), [z1 + 1, z2 + 1])
-    fg = @inferred f * g
+    fg = f * g
```
Comment (HechtiDerLachs) [inline test/Rings/MPolyAnyMap/MPolyRing.jl:127]: @thofma : I had reasons to not allow `CompositeMap`s at some points within the schemes. To achieve this, I had to overwrite composition for some constellations. But given the genererality of allowable coefficient maps, I did not see how to make this type stable anymore. Do you strongly object to this regression here?

**Comment by HechtiDerLachs** [comment]
I am sorry about the messy state that this is in. But it turned out to be a really hard nut to crack here. 

Some things to be discussed:

1) I introduced a "simultaneous blowup strategy" for the elliptic surfaces. Something which is vital to what we try to achieve these days. This should actually be consensus.

2) Because of 1) I needed to restructure the implementation of `restrict(::AbsCoveredSchemeMorhpism, ::CoveredClosedEmbedding, ::CoveredClosedEmbedding)`. This lead to a lot of issues down the road. All code which uses refinements of coverings for morphisms of covered schemes is still rather brittle. Unfortunately, this is a bigger task to clean this up. A lot has already been done, but many things still only work in very particular constellations. One thing which did not work anymore was the `isomorphism_on_open_subset` which I had once added in order to be able to identify function fields along birational maps like blowups. This feature turned out to not be that useful and we actually don't use it these days. So I took the liberty to disable it. Let's see whether the tests pass. Ping @simonbrandhorst 

3) Something went wrong with some composition of maps. I suddenly got `CompositeMap`s where earlier the composition had been carried out. I still could not pin down the original cause of this, but I made some fixes. In the course of these events, I found it necessary to move the type declaration for `MPolyAnyMap` to a file of its own. Otherwise, debugging would not have been possible. I would appreciate it, if it could stay this way. Ping @thofma .  

Once this reaches a state where tests actually pass, I will remove the code which for the time being has just been commented out. But if you have any objections about where parts of this are going, please let me know already.

**Comment by afkafkafk13** [comment]
Can you clarify what you mean by 'simultaneous blowup strategy'?

**Comment by HechtiDerLachs** [comment]
> Can you clarify what you mean by 'simultaneous blowup strategy'?

In the elliptic surfaces we exclusively encounter A-D-E singularities. Up to now we blew them up one by one. But this produces a lot of charts and gluings. In the particular example we need for our paper we only have 8 ordinary A1 singularities. They are resolved simultaneously by one single blowup. This is achieved with with a keyword argument to the constructor for elliptic surfaces which then selects this strategy: Just blow up the ideal sheaf of the reduced singular locus, until you're done.

**Comment by afkafkafk13** [comment]
> > Can you clarify what you mean by 'simultaneous blowup strategy'?
> 
> In the elliptic surfaces we exclusively encounter A-D-E singularities. Up to now we blew them up one by one. But this produces a lot of charts and gluings. In the particular example we need for our paper we only have 8 ordinary A1 singularities. They are resolved simultaneously by one single blowup. This is achieved with with a keyword argument to the constructor for elliptic surfaces which then selects this strategy: Just blow up the ideal sheaf of the reduced singular locus, until you're done.

oh. I see. This makes a lot of sense.

File: experimental/Schemes/src/BlowupMorphism.jl [hunk 12]
```diff
@@ -494,29 +532,35 @@ function restrict(f::AbsCoveredSchemeMorphism,
...
   inc_cod_cov = covering_morphism(inc_cod)
 
-  # We need to do the following.
-  # - Pass to a common refinement ref_cod in X that both
-  #   f and inc_cod can restrict to.
-  # - Pass to a common refinement in Y
-  ref_cod, a, b = _register!(common_refinement(codomain(f_cov), codomain(inc_cod_cov)), codomain(f))
-  inc_cod_ref = restrict(inc_cod, ref_cod)
-  f_res = restrict(f, ref_cod)
-  ref_dom, aa, bb = _register!(common_refinement(domain(f_res), codomain(inc_dom_cov)), domain(f))
```
Comment (simonbrandhorst) [inline experimental/Schemes/src/BlowupMorphism.jl:504]: Before the refinements were registered. Now this does not seem to be the case anymore. 
Is this intentional? And if yes what is the rationale here?

**Comment by simonbrandhorst** [comment]
I am fine with discontinuing `isomorphism_on_open_subset` it is in experimental and not exported.

**Review by simonbrandhorst — CHANGES_REQUESTED** [review]
The changes in the doctests in unrelated code such as algebraic statistics seem odd.
Can you explain the change in behavior? Are the results still correct?

File: experimental/Schemes/src/BlowupMorphism.jl (hunk 12 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/src/BlowupMorphism.jl:504]: Originally we were planning to keep track of all the refinements and the morphisms between them inside the covered scheme. But this is not very garbage-collector-friendly, so we eventually refrained from that and switched to the implicit tree structure for refinements. The call to `_register!` was a deprecated artefact.

**Comment by HechtiDerLachs** [comment]
> Can you explain the change in behavior?

See #4171 and my comment on slack.

**Comment by simonbrandhorst** [comment]
O.K. when you checkout the files I reapprove.

**Review by simonbrandhorst — APPROVED** [review]
